### ISSUE TITLE
There have been long-standing repaint-flicker issues in the libarary. At one point I thought I fixed them all with a fix to how `ClearWhenVisibleFalse` was in play.

### ISSUE THREAD

**Issue by tig** [gui-cs/Terminal.Gui#3408]
There have been long-standing repaint-flicker issues in the libarary. At one point I thought I fixed them all with a fix to how `ClearWhenVisibleFalse` was in play.

However, I did something in this PR that has caused the problem to re-appear. I've spent HOURs trying to track it down unsuccessfully. 

I am not going to hold up merging this PR to fix it, because it's cosmetic, but it needs to be found and fixed. Any help appreciated.

![FLUaXgV 1](https://github.com/gui-cs/Terminal.Gui/assets/585482/15212750-f4b5-4c80-97e7-3b39839d23b6)

Here I held down Ctrl-G to repro:

![e2DfkaP 1](https://github.com/gui-cs/Terminal.Gui/assets/585482/16748f55-1a0c-41ba-b7d2-18ad265eb411)

_Originally posted by @tig in https://github.com/gui-cs/Terminal.Gui/issues/3323#issuecomment-2057152669_

**Comment by tig** [comment]
@BDisp if you're going to look into making Clip more sophisticated, I found the PR I thought I lost where I started down that path:

https://github.com/gui-cs/Terminal.Gui/pull/2606

**Comment by BDisp** [comment]
> I'm done trying to debug this flicker thing for now. So frustrating!

That's true, it's really frustrating. I have also analyzed the cause and I am almost certain that it is necessary to limit the drawing of decorations to their thickness. It could also be someone from Label at Border who is cleaning outside of his area.

**Comment by tig** [comment]
> @BDisp if you're going to look into making Clip more sophisticated, I found the PR I thought I lost where I started down that path:
> 
> #2606

Nevermind. That branch does not have all the really awesome code I wrote in it. I must have never pushed. Long gone. Sad.

**Comment by BDisp** [comment]
> @BDisp if you're going to look into making Clip more sophisticated, I found the PR I thought I lost where I started down that path:
> 
> #2606

I've seen it, but it has a lot of unrelated changes and I'd rather start from scratch.

**Comment by tig** [comment]
> > I'm done trying to debug this flicker thing for now. So frustrating!
> 
> That's true, it's really frustrating. I have also analyzed the cause and I am almost certain that it is necessary to limit the drawing of decorations to their thickness. It could also be someone from Label at Border who is cleaning outside of his area.

Note that in #3323, someone has to directly set `Driver.Clip` to cause `Driver.AddRune` or `Driver.FIllRect` to do any drawing outside of `Viewport`. 

The only code still in the library (outside of `ViewDrawing.cs` that still does that is:

![image](https://github.com/gui-cs/Terminal.Gui/assets/585482/5cb05df8-bd6d-464c-a1e0-7ceac84106b8)

Based on this, I'm skeptical it has to do with someone clearing outside of their Viewport.

Except ScrollView:

![image](https://github.com/gui-cs/Terminal.Gui/assets/585482/bf973811-fbc8-49c9-9184-2f5a672b084d)

**Comment by BDisp** [comment]
I have already made the changes to use an array of Rectangles and the same problem continues to persist.
Using conhost directly from VS2022 this flickering does not happen, even though we are using True Color. Just using Windows Terminal this flickering continues and I already suspect if it is a problem with it.
Still, do you want me to submit a PR with the changes to Driver.Clip?

**Comment by tig** [comment]
> I have already made the changes to use an array of Rectangles and the same problem continues to persist. Using conhost directly from VS2022 this flickering does not happen, even though we are using True Color. Just using Windows Terminal this flickering continues and I already suspect if it is a problem with it. Still, do you want me to submit a PR with the changes to Driver.Clip?

I don't think an array of Rectangles is the right approach. I think we want to build the next-gen clipping support on a `Region` class that is either directly this:

https://learn.microsoft.com/en-us/dotnet/api/system.drawing.region?view=dotnet-plat-ext-8.0

Or, is a clone of it. 

So, unless you built your experiment on `Region`, no.

**Comment by BDisp** [comment]
> I don't think an array of Rectangles is the right approach. I think we want to build the next-gen clipping support on a `Region` class that is either directly this:
> 
> https://learn.microsoft.com/en-us/dotnet/api/system.drawing.region?view=dotnet-plat-ext-8.0
> 
> Or, is a clone of it.
> 
> So, unless you built your experiment on `Region`, no.

By chance I noticed the `Region` class but I didn't use it to avoid having to add another library. I created a `RectangleExtensions` class to manipulate its Rectangle methods with the array. But since it didn't solve the problem with flickering and I didn't use the `Region` class, I'm going to give up on continuing with my PR.
